[GStreamer][EME] Added support to check support of cryptoblockformat

philn opened the pull request:
#### 8f08fa82eddb19e066a11d1a0c0833008613843d
<pre>
[GStreamer][EME] Added support to check support of cryptoblockformat
<a href="https://bugs.webkit.org/show_bug.cgi?id=253000">https://bugs.webkit.org/show_bug.cgi?id=253000</a>

Reviewed by Xabier Rodriguez-Calvar.

Inspired from downstream WPE 2.22 patch:
<a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/e5b91c2bb7732e1047e3f52aa6f6651e0e1f02da">https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/e5b91c2bb7732e1047e3f52aa6f6651e0e1f02da</a>

Fixes YT cert test: &quot;Format Support Tests&quot; -&gt; &quot;1. isTypeSupported cryptoblockformat&quot;

Also drive-by change suggested by Xabier, remove the application/x-cbcs check in the Thunder
decryptor as that is now cenc with a cypher-mode.

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::isContentTypeSupported const):

Canonical link: <a href="https://commits.webkit.org/261684@main">https://commits.webkit.org/261684@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99564bb198a46ae4925da709dfbf667433362ffe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112553 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21704 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1209 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4328 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116618 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23042 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12865 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118320 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100322 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105615 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/867 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46126 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14041 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/904 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14723 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52909 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8143 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16560 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -660,6 +667,37 @@ MediaPlayerEnums::SupportsType GStreamerRegistryScanner::isContentTypeSupported(
+                return SupportsType::IsNotSupported;
+        }
+#else
+        if (!factories.hasElementForMediaType(ElementFactories::Type::Decryptor, "application/x-cenc"))
```
Shouldn't we also check for `application/x-webm-enc` as well?

philn replied:
I thought about yes, but wasn't sure what to do... When is that format selected? For matroska containers only?

calvaris replied:
Yes, WebM, Widevine only.

philn replied:
The problem is, when checking for something like `isTypeSupported('video/webm; codecs="vp09..."; cryptoblockformat=subsample')` should we return IsNotSupported when no webm-enc decryptor was found? or cenc? or neither of those? Not sure which semantics to use here.

calvaris replied:
IMHO, we should follow what [Chromium does](https://cobalt.googlesource.com/cobalt/+/refs/heads/20.lts.stable/src/starboard/shared/starboard/media/media_util.cc#138) and support only `cryptoblockformat=subsample` and only when used in webm and the WebM decryptor is present.

philn replied:
Sure but you can have vp8/vp9 encrypted in isomp4 too

calvaris replied:
Yes, but apparently Chromium supports is only for WebM so I'd keep it for that only and return `IsNotSupported` for the rest.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -104,6 +104,8 @@ GStreamerRegistryScanner::ElementFactories::ElementFactories(OptionSet<ElementFa
         rtpPayloaderFactories = gst_element_factory_list_get_elements(GST_ELEMENT_FACTORY_TYPE_PAYLOADER, GST_RANK_MARGINAL);
     if (types.contains(Type::RtpDepayloader))
         rtpDepayloaderFactories = gst_element_factory_list_get_elements(GST_ELEMENT_FACTORY_TYPE_DEPAYLOADER, GST_RANK_MARGINAL);
+    if (types.contains(Type::Decryptor))
```
I wonder if we should flag all the decryptor appearances in the scanner for `ENABLE(ENCRYPTED_MEDIA)` . I mean not only this part but almost all chunks of this patch.

philn replied:
When EME is disabled no decryptor will be registered and the factory list will be empty. I think we can live without ifdefs.

calvaris reviewed (commented):
As an optional drive-by fix, if you want, you can remove the `application/x-cbcs` check in the Thunder decryptor as that is now cenc with a `cypher-mode`.

I guess you also have to fix the build failures.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -660,6 +667,41 @@ MediaPlayerEnums::SupportsType GStreamerRegistryScanner::isContentTypeSupported(
+        if (cryptoblockformat != "subsample"_s)
+            return SupportsType::IsNotSupported;
+
+        const char* encryptionScheme = containerType.endsWith("webm"_s) ? "application/x-webm-enc" : "application/x-cenc";
```
Considering what we are commenting, I'd just bail out with `IsNotSupported` if container is not WebM then continue with the rest of the checks.

philn committed:
[GStreamer][EME] Added support to check support of cryptoblockformat
https://bugs.webkit.org/show_bug.cgi?id=253000

Reviewed by Xabier Rodriguez-Calvar.

Inspired from downstream WPE 2.22 patch:
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/e5b91c2bb7732e1047e3f52aa6f6651e0e1f02da

Fixes YT cert test: "Format Support Tests" -> "1. isTypeSupported cryptoblockformat"

Also drive-by change suggested by Xabier, remove the application/x-cbcs check in the Thunder
decryptor as that is now cenc with a cypher-mode.

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::isContentTypeSupported const):

Canonical link: https://commits.webkit.org/261684@main
